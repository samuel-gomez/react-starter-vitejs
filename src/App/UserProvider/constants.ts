export const NON_CONNECTE = 'Non ConnectÃ©';

// Used when oidc configuration is disabled
const useOidcUserMock = () => ({
  oidcUser: {
    name: 'John Doe',
    given_name: 'John',
    family_name: 'Doe',
    email: 'john.doe@test.fr',
    email_verified: true,
    website: 'test.fr',
    sub: '2',
  },
});

export default useOidcUserMock;
