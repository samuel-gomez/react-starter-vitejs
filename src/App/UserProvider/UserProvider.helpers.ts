import { useOidcUser } from '@axa-fr/react-oidc';
import isEmpty from 'lodash/isEmpty';
import { PROFILS } from 'shared/constants';
import useOidcUserMock, { NON_CONNECTE } from './constants';

type ToidcUser = {
  member_of?: string[];
  name?: string;
  axa_uid_racf?: string;
};

/**
 * getAuthName
 */

type TgetAuthName = {
  oidcUser: ToidcUser;
  isEmptyFn?: typeof isEmpty;
};

export const getAuthName = ({ oidcUser, isEmptyFn = isEmpty }: TgetAuthName) => (!isEmptyFn(oidcUser?.name ?? '') ? oidcUser.name : NON_CONNECTE);

/**
 * setAuthRole
 */
type TsetAuthRole = {
  memberOf: string;
  profils?: typeof PROFILS;
};

export const setAuthRole = ({ memberOf, profils = PROFILS }: TsetAuthRole) =>
  profils.map(profil => (memberOf.search(`${profil}`) !== -1 ? profil : '')).join('') || memberOf;

/**
 * getAuthRole
 */

type TgetAuthRole = {
  oidcUser: ToidcUser;
  setAuthRoleFn?: typeof setAuthRole;
};

export const getAuthRole = ({ oidcUser, setAuthRoleFn = setAuthRole }: TgetAuthRole) =>
  setAuthRoleFn({ memberOf: oidcUser?.member_of?.[0].replace('CN=', '') ?? '' });

/**
 * getAuthUid
 */

type TgetAuthUid = {
  oidcUser: ToidcUser;
  isEmptyFn?: typeof isEmpty;
};

export const getAuthUid = ({ oidcUser }: TgetAuthUid) => oidcUser?.axa_uid_racf ?? '';

/**
 * extractDataFromOAuthToken
 * @param {Object} oidcUser
 */

type TextractDataFromOAuthToken = {
  oidcUser: ToidcUser;
  getAuthNameFn?: typeof getAuthName;
  getAuthRoleFn?: typeof getAuthRole;
  getAuthUidFn?: typeof getAuthUid;
};

export const extractDataFromOAuthToken = ({
  oidcUser,
  getAuthNameFn = getAuthName,
  getAuthRoleFn = getAuthRole,
  getAuthUidFn = getAuthUid,
}: TextractDataFromOAuthToken) => ({
  authName: getAuthNameFn({ oidcUser }),
  authRole: getAuthRoleFn({ oidcUser }),
  authUid: getAuthUidFn({ oidcUser }),
  isLoading: !oidcUser,
});

export const getUserInfos = ({
  isEnabled,
  useOidcUserFn = useOidcUser,
  useOidcUserMockFn = useOidcUserMock,
}: {
  isEnabled?: boolean;
  useOidcUserFn?: typeof useOidcUser;
  useOidcUserMockFn?: typeof useOidcUserMock;
}) => (isEnabled ? useOidcUserFn : useOidcUserMockFn);
