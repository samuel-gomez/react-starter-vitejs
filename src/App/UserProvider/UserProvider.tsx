import { EnvironmentContext } from 'App/EnvironmentProvider';
import { createContext, useContext, useMemo, type ReactNode } from 'react';
import { extractDataFromOAuthToken, getUserInfos } from './UserProvider.helpers';

export type TUserContext = ReturnType<typeof extractDataFromOAuthToken> & {
  isLoading?: boolean;
};

export const UserContext = createContext<TUserContext>({
  authName: '',
  authRole: '',
  authUid: '',
  isLoading: true,
});
UserContext.displayName = 'UserContext';

type TUserProvider = {
  children: ReactNode;
  getUserInfosFn?: typeof getUserInfos;
  extractDataFromOAuthTokenFn?: typeof extractDataFromOAuthToken;
  isEnabled?: boolean;
};

const UserProvider = ({
  children,
  getUserInfosFn = getUserInfos,
  extractDataFromOAuthTokenFn = extractDataFromOAuthToken,
  ...rest
}: TUserProvider) => {
  const { environment } = useContext(EnvironmentContext);
  const { oidcUser } = getUserInfosFn({ isEnabled: environment?.oidc?.isEnabled })();

  const value = useMemo(() => ({ ...extractDataFromOAuthTokenFn({ oidcUser }), ...rest }), [extractDataFromOAuthTokenFn, oidcUser, rest]);
  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

export default UserProvider;
