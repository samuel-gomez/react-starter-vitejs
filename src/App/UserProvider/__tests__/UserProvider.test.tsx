import { useContext } from 'react';
import { render, screen } from '@testing-library/react';
import UserProvider, { UserContext } from '../UserProvider';

type TBase = {
  authName: string | undefined;
  authRole: string;
  authUid: string;
  isLoading?: boolean;
};

const Base = ({ authName, authRole, authUid, isLoading }: TBase) => (
  <ul>
    <li>{isLoading ? 'have isLoading' : 'notHave isLoading'}</li>
    <li>{authName ? 'have authName' : 'notHave authName'}</li>
    <li>{authRole ? 'have authRole' : 'notHave authRole'}</li>
    <li>{authUid ? 'have authUid' : 'notHave authUid'}</li>
  </ul>
);

const BaseWithUser = () => {
  const userProps = useContext(UserContext);
  return <Base {...userProps} />;
};

const oidcUser = {
  name: 'Bob Smith',
  given_name: 'Bob',
  family_name: 'Smith',
  email: 'BobSmith@email.com',
  email_verified: true,
  website: 'https://bob.com',
  sub: '11',
  member_of: ['CN=Admin'],
  axa_uid_racf: 'S000007',
};

const getUserInfosMock = vi.fn().mockReturnValue(() => ({
  oidcUser,
}));

const App = () => (
  <UserProvider getUserInfosFn={getUserInfosMock}>
    <BaseWithUser />
  </UserProvider>
);

describe('Render App with Base have user props', () => {
  it('Should Base have user props when render App with UserProvider', () => {
    render(<App />);

    expect(screen.getByText('have authName')).toBeDefined();
    expect(screen.getByText('have authRole')).toBeDefined();
    expect(screen.getByText('have authUid')).toBeDefined();
    expect(screen.getByText('notHave isLoading')).toBeDefined();
  });
});
