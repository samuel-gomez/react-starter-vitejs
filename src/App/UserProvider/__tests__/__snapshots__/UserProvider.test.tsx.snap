// Vitest Snapshot v1

exports[`Render App with Base have user props > Should Base have user props when render App with UserProvider 1`] = `
<DocumentFragment>
  <ul>
    <li>
      have isEnabled
    </li>
    <li>
      notHave isLoading
    </li>
    <li>
      have authName
    </li>
    <li>
      have authRole
    </li>
    <li>
      have authUid
    </li>
  </ul>
</DocumentFragment>
`;
